src/TSL.py

```diff
@@ -0,0 +1,224 @@
+            if max([len(x) for x in factors]) == 1:
+                warnings.warn("Spaces seem to be missing in the factor input file", SyntaxWarning)
+
+        return TSL(tier, factors)
+
+    def project(self, string):
+        '''
+            Given a list of symbols as a string, return the list with all the irrelevant symbols masked out
+        '''
+        if not string:
+            return []
+
+        first = string[0]
+        rest = string[1:]
+
+        if first in self.tier:
+            return [first] + self.project(rest)
+        else:
+            return self.project(rest)
+
+    def val(self, s):
+        '''
+            Given a string s, return True if s is accepted by the TSL grammar, False otherwise
+        '''
+        projected_string = self.project(s)
+        padded_string = tuple((self.k - 1) * ['#'] + projected_string + (self.k - 1) * ['#'])
+
+        return self.check_k_factors(padded_string)
+
+    def check_k_factors(self, s):
+        '''
+            Given a padded string s, return True if s is accepted by the TSL grammar, False otherwise
+        '''
+        if len(s) < self.k:
+            return True
+
+        k_factor = s[:self.k]
+        return not (tuple(k_factor) in self.k_factors) and self.check_k_factors(s[1:])
+
+    def eval_corpus(self, corpus_file, corpus_filename=None):
+        '''
+            Evaluate the corpus forms based on the current TSL grammar.
+
+            Input:
+                corpus_file: path to the corpus file.
+                corpus_filename: path to save corpus evaluations.
+        '''
+        corpus_data = self.read_corpus_file(corpus_file)
+        results = []
+        for string in corpus_data:
+            results.append(self.val(string))
+
+        self.eval_to_csv(corpus_data, results, corpus_filename)
+
+    def TSL_to_pTSL(self, tier_filename, factor_filename):
+        '''
+            Convert TSL grammar to pTSL and output a pTSL grammar by saving the tier and k-factor files.
+
+        '''
+        if not tier_filename:
+            filename = DEFAULT_PTSL_TIER_FILENAME
+        if not factor_filename:
+            filename = DEFAULT_PTSL_FACTOR_FILENAME
+        probs = {}
+
+        for tier in self.tier:
+            probs[tier] = 1.0
+
+        factors = self.k_factors
+
+        with open(tier_filename, 'w', newline="") as f:
+            csv_writer = csv.writer(f)
+            for key, value in probs.items():
+                csv_writer.writerow([key, value])
+
+        with open(factor_filename, 'w', newline="") as f:
+            csv_writer = csv.writer(f)
+            for sym in factors:
+                csv_writer.writerow(sym)
+
+    @staticmethod
+    def eval_to_csv(corpus_data, results, filename=None):
+        '''
+            Save the result of eval to output filename
+        '''
+        if not filename:
+            filename = DEFAULT_EVAL_FILENAME
+        with open(filename, 'w', newline="") as f:
```

Comment:
I research a little bit and found this explanation on the official documentation:

newline controls how universal newlines mode works (it only applies to text mode). It can be None, '', '\n', '\r', and '\r\n'. It works as follows:

When reading input from the stream, if newline is None, universal newlines mode is enabled. Lines in the input can end in '\n', '\r', or '\r\n', and these are translated into '\n' before being returned to the caller. If it is '', universal newlines mode is enabled, but line endings are returned to the caller untranslated. If it has any of the other legal values, input lines are only terminated by the given string, and the line ending is returned to the caller untranslated.

When writing output to the stream, if newline is None, any '\n' characters written are translated to the system default line separator, os.linesep. If newline is '' or '\n', no translation takes place. If newline is any of the other legal values, any '\n' characters written are translated to the given string.

So either works in our case, but most implementation I saw has it as '' instead of None as a default.